src/lib/updating-element.ts

```diff
@@ -475,85 +465,79 @@ export abstract class UpdatingElement extends HTMLElement {
-      if (!(this._validationState & STATE_HAS_UPDATED)) {
-        // mark state has updated
-        this._validationState = this._validationState | STATE_HAS_UPDATED;
-        if (typeof this.finishFirstUpdate === 'function') {
-          // During `finishFirstUpdate` (which is optional), setting properties triggers invalidation,
-          // and users may choose to await other state.
-          const result = this.finishFirstUpdate();
-          if (result != null && typeof (result as PromiseLike<unknown>).then === 'function') {
-            await result;
-          }
-        }
-      }
-      // During `finishUpdate` (which is optional), setting properties triggers invalidation,
-      // and users may choose to await other state, like children being updated.
-      if (hasFinishUpdate) {
-        const result = this.finishUpdate!(changedProperties as PropertyValues);
-        if (result != null && typeof (result as PromiseLike<unknown>).then === 'function') {
-          await result;
-        }
+    if (!this.shouldUpdate(this._changedProperties)) {
+      this._markUpdated();
+      return;
+    }
+    // During update, setting properties does not trigger invalidation.
+    this.update(this._changedProperties);
+    // copy changedProperties to hand to finishUpdate.
+    const changedProperties = this._changedProperties;
+    this._markUpdated();
+    // After update (finishFirstUpdate, finishUpdate), properties *do* trigger invalidation.
+    if (!(this._validationState & STATE_HAS_UPDATED)) {
+      // mark state has updated
+      this._validationState = this._validationState | STATE_HAS_UPDATED;
+      if (typeof this.finishFirstUpdate === 'function') {
+        this.finishFirstUpdate();
       }
-    } else {
-      this._changedProperties.clear();
-      // mark state valid
-      this._validationState = this._validationState | STATE_IS_VALID;
     }
-    // Only resolve the promise if we finish in a valid state (finishUpdate
-    // did not trigger more work). Note, if invalidate is triggered multiple
-    // times in `finishUpdate`, only the first time will resolve the promise
-    // by calling `_validateResolver`. This is why we guard for its existence.
-    if ((this._validationState & STATE_IS_VALID) && typeof this._validateResolver === 'function') {
-      this._validateResolver();
-      this._validateResolver = undefined;
+    if (typeof this.finishUpdate === 'function') {
+      this.finishUpdate(changedProperties);
     }
-    return this._validatePromise;
   }
 
-  private get _isPendingUpdate() {
-    return !(this._validationState & STATE_IS_VALID);
+  private _markUpdated() {
+    this._changedProperties = new Map();
+    this._validationState = this._validationState & ~STATE_IS_UPDATING;
   }
 
   /**
-   * Returns a Promise that resolves when the element has finished updating.
+   * Returns a Promise that resolves when the element has finished updating
```

Comment:
it "resolves to a boolean value" so I would use commas around the injected sentence to make it more readable

Or maybe rewrite (maybe this is more clear)

Returns a Promise that resolves to a boolean when the element has finished updating. The boolean result value is true if there are no further pending updates. New pending updates can eg. be the result of a property being set in `finishUpdate`.